File: solutions/979.di-string-match/di-string-match.java
class Solution {
    public int[] diStringMatch(String S) {
        int n = S.length(), left = 0, right = n;
        int[] res = new int[n + 1];
        int i = 0;
        for(char c : S.toCharArray()) {
            if(c == 'I') res[i++] = left++;
            else res[i++] = right--;
        }
        res[n] = left;
        return res;
    }
}

File: solutions/979.di-string-match/di-string-match.cpp
class Solution {
public:
    vector<int> diStringMatch(string S) {
        int left = count(S.begin(), S.end(), 'D'), right = left;
        vector<int> res = {left};
        for(char c : S) {
            if(c == 'I') res.push_back(++right);
            else res.push_back(--left);
        }
        return res;
    }
};